src-tauri/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+            .to_bytes()
+            .unwrap(),
+    )
+    .unwrap();
+    let token = totp.generate_current().unwrap();
+
+    // Step 4: Enter TOTP token
+    navigated_tab.wait_for_element("input#userid");
+    navigated_tab
+        .find_element("input#userid")?
+        .type_into(token.as_str())?;
+
+    // Submit the form
+    navigated_tab
+        .find_element("button.button-orange.wide")?
+        .click()?;
+    navigated_tab.wait_for_element("strong#request_token");
+
+    let requested_token = navigated_tab
+        .find_element("strong#request_token")?
+        .get_inner_text()?;
+
+    // Step 5: Retrieve the token from localStorage
+    // Define the JavaScript to retrieve the token from localStorage
+    let script = r#"
+     (function() {
+         return localStorage.getItem('__storejs_dpmorgan_enctoken');
+     })();
+ "#; // Evaluate the JavaScript in the context of the page
+    let encoded_token = navigated_tab
+        .evaluate(script, true)?
+        .value
+        .unwrap()
+        .to_string();
+
+    // Check if the token exists
+    if !requested_token.is_empty() {
+        if !encoded_token.is_empty() {
+            // Check if the token matches the expected format
+            //let encoded_token = encoded_token.trim();
+            let regex = Regex::new(r"^.{1,}$").unwrap(); // Assuming TOTP tokens are 21-character strings
+
+            // Match the token against the regex
+            if regex.is_match(encoded_token.clone().as_str()) {
+                return Ok((requested_token, encoded_token));
+            } else {
+                // Returning the error
+                return Err(Box::new(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "Invalid encoded token found in the localStorage",
+                )));
+            }
+        } else {
+            // Returning the error
+            return Err(Box::new(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "No encoded token found in the localStorage",
+            )));
+        }
+    } else {
+        // Returning the error
+        return Err(Box::new(std::io::Error::new(
+            std::io::ErrorKind::Other,
+            "Could not find the requested token in the page",
+        )));
+    }
+    return Ok((String::new(), String::new()));
+}
+
+pub async fn handle_margin(
+    browser: Browser,
+    margin_amount: i32,
+) -> Result<i32, Box<dyn std::error::Error>> {
+    const MARGIN_URL: &str =
+        "http://ec2-3-110-151-1.ap-south-1.compute.amazonaws.com/tmp/simulate/margin.html";
+
+    const LOGIN_URL: &str =
+        "http://ec2-3-110-151-1.ap-south-1.compute.amazonaws.com/tmp/simulate/login.html";
+    const WAITING_TIME: u64 = 3; //s
+
+    // Step 6: Find the request token tab and open the popup
+    for tab in browser.get_tabs().lock().unwrap().iter() {
+        if tab.get_url().contains("simulate/session.html") {
```

Comment:
Are we trying to identify the window to re-use for add margin? I don't seem to understand this part.

Reply:
Yes, we are investigating the open tabs to find the adequate popup window. I could not find a different solution than this filtering.